patches/api/0374-Add-Recipe-API.patch

```diff
@@ -0,0 +1,273 @@
++public class RecipeType<R extends Recipe, I extends Inventory> implements Keyed, org.bukkit.Keyed {
++    private final @NotNull NamespacedKey key;
++
++    public RecipeType(@NotNull String key) {
++        this.key = Objects.requireNonNull(org.bukkit.NamespacedKey.fromString(key));
++    }
++
++    public RecipeType(@NotNull NamespacedKey key) {
++        this.key = Preconditions.checkNotNull(key, "key");
++    }
++
++    @Override
++    public @NotNull NamespacedKey getKey() {
++        return key;
++    }
++
++    @Override
++    public @NotNull Key key() {
++        return key;
++    }
++
++    private static final Map<NamespacedKey, RecipeType<?, ?>> recipeTypes = new HashMap<>(7);
++
++    private static <R extends Recipe, C extends Inventory> RecipeType<R, C> register(final @NotNull String id) {
++        RecipeType<R, C> recipeType = new RecipeType<>(id);
++        recipeTypes.put(recipeType.key, recipeType);
++        return recipeType;
++    }
++
++    /**
++     * Recipes for crafting in a {@link org.bukkit.Material#CRAFTING_TABLE CRAFTING_TABLE}.
++     */
++    public static final RecipeType<Recipe, CraftingInventory> CRAFTING = register("crafting");
++    /**
++     * Recipes for smelting inside of a {@link org.bukkit.Material#FURNACE FURNACE}.
++     */
++    public static final RecipeType<FurnaceRecipe, Inventory> SMELTING = register("smelting");
++    /**
++     * Recipes for smelting inside of a {@link org.bukkit.Material#BLAST_FURNACE BLAST_FURNACE}.
++     */
++    public static final RecipeType<BlastingRecipe, Inventory> BLASTING = register("blasting");
++    /**
++     * Recipes for smelting inside of a {@link org.bukkit.Material#SMOKER SMOKER}.
++     */
++    public static final RecipeType<SmokingRecipe, Inventory> SMOKING = register("smoking");
++    /**
++     * Recipes for smelting on a {@link org.bukkit.Material#CAMPFIRE CAMPFIRE}.
++     */
++    public static final RecipeType<CampfireRecipe, Inventory> CAMPFIRE_COOKING = register("campfire_cooking");
++    /**
++     * Recipes for cutting in a {@link org.bukkit.Material#STONECUTTER STONECUTTER}.
++     */
++    public static final RecipeType<StonecuttingRecipe, Inventory> STONECUTTING = register("stonecutting");
++    /**
++     * Recipes for upgrading items in a {@link org.bukkit.Material#SMITHING_TABLE SMITHING_TABLE}.
++     */
++    public static final RecipeType<SmithingRecipe, Inventory> SMITHING = register("smithing");
++
++    /**
++     * Gets a registered recipe type by its key.
++     * @return a registered recipe type or null
++     */
++    public static @Nullable RecipeType<?, ?> getByKey(final @NotNull NamespacedKey key) {
++        return recipeTypes.get(Preconditions.checkNotNull(key, "key"));
++    }
++
++    /**
++     * Gets an iterator across all registered recipe types.
++     * @return recipe type iterator
++     */
++    public static @NotNull Iterator<RecipeType<?, ?>> iterator() {
++        return recipeTypes.values().iterator();
++    }
```

Comment:
This should be moved over to a server-backed registry via the server PaperRegistry and UnsafeValues#registryFor.

Look at `ConfiguredStructure` for reference.

Reply:
Added a pretty naive implementation of this; hope it's what you were looking for ^^

Reply:
looks good. when making that I totally didn't consider classes with generic params. so you had to do that weird double casting thing instead of using like a type token. That should be addressed before, but it's small.

Reply:
yea type tokens would be a good idea. also oh my i'm not sure what to do about those unit tests failing 😬